### PR TITLE
Fix cppcheck warnings

### PR DESCRIPTION
**Pull request by tehnick**

### CODE REVIEW

File: examples/GuiClient/chatGraphicsScene.cpp [hunk 1]
```diff
@@ -27,7 +27,7 @@
 #include "chatGraphicsView.h"
 
 chatGraphicsScene::chatGraphicsScene(QObject* parent) : QGraphicsScene(parent),
-        m_verticalPosForNewMessage(0), m_verticalSpacing(5)
+        m_verticalPosForNewMessage(0), m_verticalSpacing(5), m_boxStartLength(5)
```
Comment (0xd34df00d) [inline examples/GuiClient/chatGraphicsScene.cpp]: Where does this value come from?

File: src/base/QXmppJingleIq.cpp [hunk 2]
```diff
@@ -710,7 +710,7 @@ class QXmppJingleIqPrivate : public QSharedData
 };
 
 QXmppJingleIqPrivate::QXmppJingleIqPrivate()
-    : ringing(false)
+    : action(QXmppJingleIq::ContentAccept), ringing(false)
```
Comment (0xd34df00d) [inline src/base/QXmppJingleIq.cpp]: ditto

File: src/base/QXmppRtpChannel.h [hunk 3]
```diff
@@ -209,6 +209,9 @@ class QXMPP_EXPORT QXmppVideoFrame
 class QXMPP_EXPORT QXmppVideoFormat
 {
 public:
+    QXmppVideoFormat() : m_frameRate(15.0), m_frameSize(QSize(320, 240)),
+        m_pixelFormat(QXmppVideoFrame::Format_YUYV) {}
```
Comment (0xd34df00d) [inline src/base/QXmppRtpChannel.h]: ditto

File: src/client/QXmppCallManager.cpp [hunk 4]
```diff
@@ -104,7 +104,8 @@ class QXmppCallManagerPrivate
 };
 
 QXmppCallPrivate::QXmppCallPrivate(QXmppCall *qq)
-    : manager(0),
+    : direction(QXmppCall::IncomingDirection),
```
Comment (0xd34df00d) [inline src/client/QXmppCallManager.cpp]: ditto

File: src/base/QXmppElement.cpp [hunk 5]
```diff
@@ -123,10 +123,13 @@ QXmppElement::~QXmppElement()
 
 QXmppElement &QXmppElement::operator=(const QXmppElement &other)
 {
-    other.d->counter.ref();
-    if (!d->counter.deref())
-        delete d;
-    d = other.d;
+    if (this != &other) // self-assignment check
```
Comment (0xd34df00d) [inline src/base/QXmppElement.cpp]: This check doesn't actually affect correctness, right?

If so, I'm not sure it should be there. The performance improvements might be marginal, if any. Also, note that, say, typical std::string implementations don't do any self-assignment checks in their operator= (at least, I'm not aware of any implementations doing so).

File: src/base/QXmppRtpChannel.h (hunk 3 above)
Comment (tehnick) [inline src/base/QXmppRtpChannel.h]: https://github.com/qxmpp-project/qxmpp/blob/034c318/src/base/QXmppRtpChannel.cpp#L809

File: src/client/QXmppTransferManager.cpp [hunk 6]
```diff
@@ -220,6 +220,7 @@ class QXmppTransferJobPrivate
 QXmppTransferJobPrivate::QXmppTransferJobPrivate()
     : blockSize(16384),
     client(0),
+    direction(QXmppTransferJob::IncomingDirection),
```
Comment (tehnick) [inline src/client/QXmppTransferManager.cpp]: First element in enum.

File: src/client/QXmppCallManager.cpp (hunk 4 above)
Comment (tehnick) [inline src/client/QXmppCallManager.cpp]: First element in enum.

File: src/base/QXmppJingleIq.cpp (hunk 2 above)
Comment (tehnick) [inline src/base/QXmppJingleIq.cpp]: First element in enum.

File: examples/GuiClient/chatGraphicsScene.cpp (hunk 1 above)
Comment (tehnick) [inline examples/GuiClient/chatGraphicsScene.cpp]: From nowhere. The main idea was that it should not be equal to null.

File: src/base/QXmppElement.cpp (hunk 5 above)
Comment (tehnick) [inline src/base/QXmppElement.cpp]: > This check doesn't actually affect correctness, right?

Yep.

> The performance improvements might be marginal

Are you sure? This looks quite long:
https://github.com/qxmpp-project/qxmpp/blob/034c318/src/base/QXmppElement.cpp#L53

Comment (tehnick) [inline src/base/QXmppElement.cpp]: > This check doesn't actually affect correctness, right?

Yep.

> The performance improvements might be marginal, if any.

I agree.

> If so, I'm not sure it should be there. 

So am I. But I also do not see reasons why not.

File: src/base/QXmppRtpChannel.h (hunk 3 above)
Comment (jlaine) [inline src/base/QXmppRtpChannel.h]: On a stylistic point, let's please align with Qt:

class Foo
    : foo(0)
    , bar(1)

File: src/base/QXmppElement.cpp (hunk 5 above)
Comment (jlaine) [inline src/base/QXmppElement.cpp]: Qt does no self assignment checks either

Comment (tehnick) [inline src/base/QXmppElement.cpp]: 1) Qt uses smart pointers (QSharedDataPointer, etc) in many classes.
2) There are few places in Qt there self assignment checks are present. For example:
http://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/tools/qcollator.cpp?h=v5.8.0#n107
http://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/io/qdebug.h?h=v5.8.0#n200
http://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qvariant.cpp?h=v5.8.0#n1881
http://code.qt.io/cgit/qt/qtbase.git/tree/src/gui/text/qtextoption.cpp?h=v5.8.0#n122
etc.

**Comment by tehnick** [comment]
> I am a little concerned by vast sweeping changes which don't seem to correct any actual bugs.

Do you think it is better to leave uninitialized variables in the *library*? Even if they will not cause segfaults due to specifics of usage of these classes, it is bad idea IMHO.

**Comment by tehnick** [comment]
> On a stylistic point, let's please align with Qt:

Done.

File: src/base/QXmppElement.cpp (hunk 5 above)
Comment (tehnick) [inline src/base/QXmppElement.cpp]: > Also, note that, say, typical std::string implementations don't do any self-assignment checks in their operator= (at least, I'm not aware of any implementations doing so).

Why have you mentioned `std::string`? I do not know about it, but other STL classes do have self-assignment checks in `operator=`. At least in libstdc++.

Comment (tehnick) [inline src/base/QXmppElement.cpp]: And of course I may remove this commit from PR if you are against it.

**Comment by tehnick** [comment]
Rebased against master.

Any additional comments?

**Comment by tehnick** [comment]
Any updates on this thread?

**Comment by tehnick** [comment]
Please merge or reject.

**Comment by 0xd34df00d** [comment]
@jlaine while I agree vast changes might introduce bugs, fixing even false positives from tools like cppcheck (even though they are IMO crappy compared to clang-tools/coverity) reduces the noise on subsequent runs and allows spotting the actual detected bugs more easily.

So, I'd like to merge this if you don't mind.

**Comment by jlaine** [comment]
If you could just add a test case for the self assignment you're fixing, go ahead and merge Boris.

**Comment by tehnick** [comment]
PR is rebased against current master.

**Comment by tehnick** [comment]
cppcheck logs for current master:
https://tehnick.net/sibuserv/projects/qxmpp/v0.9.3-49-g09560c9/StaticCodeAnalysis/cppcheck.log
https://tehnick.net/sibuserv/projects/qxmpp/v0.9.3-49-g09560c9/StaticCodeAnalysis/cppcheck.html/

**Pull request merged**